### PR TITLE
Pull Request 2

### PR DESCRIPTION
**Pull request by dandiesta**

### CODE REVIEW

File: app/controllers/comment_controller.php [hunk 1]
```diff
@@ -63,26 +66,103 @@ public function edit()
...
-                $page = 'edit';
-            }
-            break;  
-        default:
-            throw new NotFoundException("{$page} not found");
-            break;
+            case 'edit':
+                break;
+            case 'edit_end':
+               $comment->body = Param::get('body');
```
Comment (romeoartista) [inline app/controllers/comment_controller.php:101]: Same as [this comment](https://github.com/erikjohnmoreno/board_exercises/pull/2#discussion_r24557679).

File: app/controllers/comment_controller.php [hunk 2]
```diff
@@ -63,26 +66,103 @@ public function edit()
...
+        $comment = new Comment();
+
+        $comment_id = Param::get('comment_id');
+        $like_checker = $comment->hasLiked($comment_id);
+        $dislike_checker = $comment->hasDisliked($comment_id);
+
+        if (!$like_checker) { //comment has not been liked
+            if (!$dislike_checker) { //comment has not been disliked
+                $comment->addLike($comment_id); //add 1 liked record in like_monitor table
+                $comment->updateLikedCount($comment_id); //increment liked column in comment table
```
Comment (charmgranada) [inline app/controllers/comment_controller.php]: Is it possible to just put the same logic under 1 function vs using 2 separate functions (addLike, updateLikedCount)? I think if 1 function fails, the other should not proceed as well po :bow: 
Or possible to just call "updateLikedCount" inside the "addLike" function?

File: app/controllers/comment_controller.php [hunk 3]
```diff
@@ -63,26 +66,103 @@ public function edit()
...
+            } else { //comment has been disliked
+                $comment->subtractDislikedCount($comment_id); //subtract 1 from disliked column in comment table
+                $comment->deleteExisting($comment_id); //deletes disliked record in like_monitor table
```
Comment (charmgranada) [inline app/controllers/comment_controller.php]: same with  (addLike, updateLikedCount) comment: https://github.com/dandiesta/board_exercise/pull/2/files#r24567267

File: app/models/comment.php [hunk 4]
```diff
@@ -75,4 +144,165 @@ public function edit($comment_id)
...
+            $delete = $db->query('DELETE FROM comment WHERE thread_id = ?', array($thread_id));
+
+            $db->commit();
+        } catch (Exception $e) {
+            $db->rollback();
+        }
+    }
+
+    //delete one comment
+    public function deleteComment($comment_id)
```
Comment (charmgranada) [inline app/models/comment.php]: Since this class is class is already named "Comment", you can remove redundancy by renaming this function to "delete", and renaming the deleteComments to "deleteAll" :bow:

File: app/models/comment.php [hunk 5]
```diff
@@ -75,4 +144,165 @@ public function edit($comment_id)
...
+    //decrements disliked column in comment table
+    public function subtractDislikedCount($comment_id)
+    {
+        $comment = new Comment();
+        
+        try {
+            $db = DB::conn();
+            $db->begin();
+
+            $update = $db->query('UPDATE comment SET disliked = disliked - 1 WHERE id = ?', array($comment_id));
```
Comment (charmgranada) [inline app/models/comment.php:208]: (low priority) I think it's cleaner to read:

``````
... SET disliked -= 1 WHERE ....
``` :bow: 
``````

File: app/models/user.php [hunk 6]
```diff
@@ -78,39 +107,77 @@ public static function get()
...
+    {
+        $db = DB::conn();
+
+        $params = array(
+            'username' => $this->username, 
+            'email'    => $this->email
+        );
+
+        $is_banned = $db->value('SELECT id FROM user WHERE username = :username AND status = 2 || 
+            email = :email AND status = 2', $params);
```
Comment (charmgranada) [inline app/models/user.php]: Please avoid magic numbers for the "status = 1" or "status 2".
You can define a const variable for it so that it's descriptive :bow:

File: app/models/user.php [hunk 7]
```diff
@@ -78,39 +107,77 @@ public static function get()
...
+
+        return (!$is_banned); //return true
```
Comment (charmgranada) [inline app/models/user.php]: The return value and the comment is a bit misleading :sweat_smile: 
You can improve on the comment statement to make it easier to understand po :bow:

File: app/controllers/comment_controller.php (hunk 2 above)
Comment (dandiesta) [inline app/controllers/comment_controller.php]: I'll just call the method updateLikedCount inside addLike and put it inside a transaction po.

File: app/models/comment.php (hunk 4 above)
Comment (dandiesta) [inline app/models/comment.php]: noted po.

File: app/models/user.php [hunk 8]
```diff
@@ -184,4 +269,86 @@ public function memberSince()
...
+    public function editStatus()
+    {
+        try {
+            $db = DB::conn();
+            $db->begin();
+
+            if ($this->current_status == 1) {
+                $update = $db->query('UPDATE user SET status= 2 WHERE id=:id', array('id' => $this->user_id));
+            } else {
+                $update = $db->query('UPDATE user SET status= 1 WHERE id=:id', array('id' => $this->user_id));
```
Comment (charmgranada) [inline app/models/user.php]: same comment: https://github.com/dandiesta/board_exercise/pull/2/files#r24567566

File: app/models/user.php [hunk 9]
```diff
@@ -184,4 +269,86 @@ public function memberSince()
...
+    //gets top commentors of the user
+    public function topCommentors()
+    {
+        $users = array();
+        $db = DB::conn();
+
+        $rows = $db->rows('SELECT a.username AS Commentor, COUNT(c.id) AS Number_of_comments FROM comment c 
+            INNER join thread t on t.id=c.thread_id 
+            INNER join user u on u.id=t.user_id 
+            INNER JOIN user a on a.id=c.user_id
```
Comment (charmgranada) [inline app/models/user.php:340]: Please avoid using Joins :bow:

File: app/models/comment.php (hunk 5 above)
Comment (dandiesta) [inline app/models/comment.php:208]: done po.

File: app/models/user.php (hunk 6 above)
Comment (dandiesta) [inline app/models/user.php]: noted po.

File: app/models/user.php (hunk 9 above)
Comment (dandiesta) [inline app/models/user.php:340]: will refactor this after all revisions because it will take me time

File: app/views/layouts/default.php [hunk 10]
```diff
@@ -6,7 +6,7 @@
     
     
     <link href="/bootstrap/css/bootstrap.min.css" rel="stylesheet">
-    
+    <link href="/bootstrap/css/bootstrap.css" rel="stylesheet">
```
Comment (bryanklab) [inline app/views/layouts/default.php]: you already linked the minified css, no need to linked the "unminified" one.

File: app/views/user/profile.php [hunk 11]
```diff
@@ -1,21 +1,92 @@
...
     <input class="span6" type="text" value="<?php echo $member_since ?> ago" name="member_since" disabled>
     <br/>
      <input type="hidden" name="page_next" value="success_update">
     <button type="submit" class="btn btn-danger">Save</button>
+    <a href="<?php enquote_string(url('user/change_password'))?>" class="btn btn-danger">Change Password</a>
 </form>
 
-<form class="offset1 span3 well">
-    Insert avatar here
+<form class="offset1 span3 well" style="box-shadow: 10px 10px 10px #888888">
```
Comment (bryanklab) [inline app/views/user/profile.php]: this is only a suggestion, you could've create your own css file then put all your inline css in that, in that way you could reuse the styles and make the code neat.

File: .gitignore [hunk 12]
```diff
@@ -1,2 +1,3 @@
 app/config/core.php
 app/tmp/logs/*.log
+app/views/*
```
Comment (fatimaPHP) [inline .gitignore]: please remove this po :bow:

File: app/app_controller.php [hunk 13]
```diff
@@ -4,18 +4,35 @@ class AppController extends Controller
...
+        if (in_array(Param::get(DC_ACTION), $exclude)) return;
 
-		if (!isset($_SESSION['userid'])) {
-		    redirect('/user/login');
-		}
-	}
+        if (!isset($_SESSION['userid'])) {
+            redirect('/user/login');
+        } else {
+            $appmodel = new AppModel();
```
Comment (fatimaPHP) [inline app/app_controller.php]: rather than modifying the `AppModel`, I think it's better to re-use the existing method in User model po :bow:

``` php
$user = User::get($user_id);
if ($user->status === ACTIVE) {
```

File: app/controllers/comment_controller.php [hunk 14]
```diff
@@ -63,26 +80,102 @@ public function edit()
...
-            throw new NotFoundException("{$page} not found");
-            break;
+            case 'edit':
+                break;
+            case 'edit_end':
+               $comment->body = Param::get('body');
+
+                try {
+                    $comment->edit($comment_id);
+                    redirect("/comment/view?thread_id={$_SESSION['thread_id']}");
```
Comment (fatimaPHP) [inline app/controllers/comment_controller.php]: please use the router po :bow:

File: app/controllers/comment_controller.php [hunk 15]
```diff
@@ -63,26 +80,102 @@ public function edit()
...
+    public function delete()
+    {
+        $comments = new Comment();
+
+        $comment_id = Param::get('comment_id');
+
+        $comments->delete($comment_id);
+
+        $this->set(get_defined_vars());
+        redirect("/comment/view?thread_id={$_SESSION['thread_id']}");
```
Comment (fatimaPHP) [inline app/controllers/comment_controller.php]: please use the router here as well and in all other areas like this po :bow:

File: app/controllers/comment_controller.php [hunk 16]
```diff
@@ -63,26 +80,102 @@ public function edit()
...
         $this->render($page);
     }
+
+    public function delete()
+    {
+        $comments = new Comment();
+
+        $comment_id = Param::get('comment_id');
+
+        $comments->delete($comment_id);
```
Comment (fatimaPHP) [inline app/controllers/comment_controller.php]: Please check if the comment is existing first before calling the `delete()` method. :bow:
This is to get the object `Comment` po, rather than initializing it.. ie.

``` php
try {
    $comment = Comment::get($comment_id);
    $comment->delete();
} catch (exception) {
    // render page if comment is not found
}
```

Also, please add the proper restriction po, anyone who can access `commend/delete?comment_id=` can delete any comment, even if they are not logged in or the comment is not created by them. :bow:

File: app/controllers/comment_controller.php [hunk 17]
```diff
@@ -63,26 +80,102 @@ public function edit()
...
+        $this->set(get_defined_vars());
+        redirect("/comment/view?page={$_SESSION['current_page']}&thread_id={$_SESSION['thread_id']}");
+    }
+
+    public function most_liked()
+    {
+        $comments = new Comment();
+        $users = new User();
+
+        $comment = $comments->getTopComments();
```
Comment (fatimaPHP) [inline app/controllers/comment_controller.php:197]: make this a static method so you do not have to initialized `Comment` po :smile:

File: app/controllers/comment_controller.php [hunk 18]
```diff
@@ -63,26 +80,102 @@ public function edit()
...
+        $user = $users->getAll();
+
+        if ($comment) {
+            $current_page = max(Param::get('page'), SimplePagination::MIN_PAGE_NUM);
+            $pagination = new SimplePagination($current_page, self::MAX_ITEMS_PER_PAGE);
+            $other_comments = array_slice($comment, $pagination->start_index + SimplePagination::MIN_PAGE_NUM);
+            $pagination->checkLastPage($other_comments);
+            $page_links = Pagination(count($comment), self::MAX_ITEMS_PER_PAGE, $current_page, self::ADJACENT_TO_CURRENT);
+            $comment = array_slice($comment, $pagination->start_index - 1, $pagination->count);
+        }
```
Comment (fatimaPHP) [inline app/controllers/comment_controller.php:208]: kindly check this [guide](http://goo.gl/R7giR0) po on how to use `SimplePagination`, so you can add the offset and limit on the static method I mentioned in L166 :bow:

File: app/controllers/thread_controller.php [hunk 19]
```diff
@@ -4,39 +4,46 @@ class ThreadController extends AppController
 {
     const MAX_ITEMS_PER_PAGE = 5;
     const MIN_PAGE_NUMBER = 1;
+    const ADJACENT_TO_CURRENT = 4;
 
     public function index()
     {
         $threads = Thread::getAll();
+        $users = new User();
+        $user = $users->getAll();
```
Comment (fatimaPHP) [inline app/controllers/thread_controller.php]: same as my comment [here](https://github.com/dandiesta/board_exercise/pull/2/files#r24880875) and [here](https://github.com/dandiesta/board_exercise/pull/2/files#r24880965) po :smile:

File: vendor/SimplePagination.php [hunk 20]
```diff
@@ -11,18 +11,29 @@
  */
 class SimplePagination
 {
-    public $current_page;        // 現在のページ番号
+    const MIN_PAGE_NUM = 1;
```
Comment (fatimaPHP) [inline vendor/SimplePagination.php]: please remove modification in this library po :bow:

File: app/controllers/comment_controller.php [hunk 21]
```diff
@@ -63,26 +80,102 @@ public function edit()
...
+    public function most_liked()
+    {
+        $comments = new Comment();
+        $users = new User();
+
+        $comment = $comments->getTopComments();
+        $user = $users->getAll();
+
+        if ($comment) {
+            $current_page = max(Param::get('page'), SimplePagination::MIN_PAGE_NUM);
```
Comment (fatimaPHP) [inline app/controllers/comment_controller.php]:
``` php
Param::get('page', 1);
```

you can do this so that 1 will be the default page if page is not specified :smile:

File: app/controllers/user_controller.php [hunk 22]
```diff
@@ -11,12 +15,12 @@ public function registration()
...
-            $register->firstname = Param::get('firstname');
-            $register->lastname = Param::get('lastname');
-            $register->username = Param::get('username');
+            $register->firstname = trim(Param::get('firstname'));
+            $register->lastname = trim(Param::get('lastname'));
+            $register->username = trim(Param::get('username'));
             $register->password = Param::get('password');
             $register->confirm_password = Param::get('confirm_password');
-            $register->email = Param::get('email');
+            $register->email = trim(Param::get('email'));
```
Comment (fatimaPHP) [inline app/controllers/user_controller.php]: suggestion po, you can also do this for readability :smile:

``` php
$params = array(
    'firstname' => Param::get('firstname'),
    'lastname' => Param::get('lastname'),
    ...
);

$user = new User($params);
$user->add();
```

File: app/models/user.php [hunk 23]
```diff
@@ -60,15 +62,40 @@ class User extends AppModel
...
+        }
+
+        return $rows;   
+    }
+
+    public static function get($user_id = null)
+    {
+        $db= DB::conn();
+        if ($user_id == null) {
+            $row = $db->row('SELECT * FROM user WHERE id=?', array($_SESSION['userid']));
```
Comment (fatimaPHP) [inline app/models/user.php]: I think session is used only in controller or view po, not in model :sweat:

File: app/models/comment.php [hunk 24]
```diff
@@ -75,4 +140,214 @@ public function edit($comment_id)
...
+        } catch (Exception $e) {
+            $db->rollback();
+        }
+    }
+
+    public function count($user_id)
+    {
+        $db = DB::conn();
+
+        $count_comments = $db->value('SELECT COUNT(id) FROM comment WHERE user_id = ?', array($user_id));
```
Comment (fatimaPHP) [inline app/models/comment.php:220]: you may directly return the `$db` result po :bow:

File: app/models/comment.php [hunk 25]
```diff
@@ -75,4 +140,214 @@ public function edit($comment_id)
...
+        return $count_likes;
+    }
+
+    public function deleteLike($thread_id)
+    {
+        try {
+            $db = DB::conn();
+            $db->begin();
+
+            $delete = $db->query('DELETE l FROM like_monitor l
```
Comment (fatimaPHP) [inline app/models/comment.php]: You may use 3 delete queries po instead of 1 query since this is inside a transaction :bow:

File: app/models/thread.php [hunk 26]
```diff
@@ -80,12 +78,18 @@ public function create(Comment $comment)
...
             $db = DB::conn();
             $db->begin();
 
-            $update = $db->query('UPDATE thread SET last_modified=now() WHERE id= ?', array($thread_id));
+            $params = array(
+                'time'      => $current_time, 
+                'thread_id' => $thread_id
+            );
+
+            $update = $db->query('UPDATE thread SET last_modified = :time WHERE id = :thread_id', $params);
```
Comment (fatimaPHP) [inline app/models/thread.php]: `$db->query` does not have any return po, please remove the extra variable :sweat:

File: app/models/user.php [hunk 27]
```diff
@@ -78,39 +105,92 @@ public static function get()
         return $row;
     }
 
+    /* CHECKERS FOR REGISTRATION */
+
+    //checks if firstname and lastname contain letters only
     public function nameChecker($name)
```
Comment (fatimaPHP) [inline app/models/user.php]: this should be in the helper po as well as L117 :sweat:

File: app/models/user.php [hunk 28]
```diff
@@ -184,4 +301,134 @@ public function memberSince()
...
+
+        foreach ($rows as $row) {
+             $users[] = new User($row);
+        }
+
+        return $users;
+    }
+
+    // Encrypt Function
+    public function mc_encrypt($encrypt, $key)
```
Comment (fatimaPHP) [inline app/models/user.php]: encrypt/decrypt method seems like a helper function, please remove it from this class as this is not related to the user :bow:

File: app/app_controller.php [hunk 29]
```diff
@@ -4,18 +4,35 @@ class AppController extends Controller
...
+
+            if ($checker == ACTIVE) {
+                return;
+            } else {
+                redirect('/user/banned');
+            }
+        }
+    }
+
+    function banned()
```
Comment (fatimaPHP) [inline app/app_controller.php:30]: is this used po ?

File: app/controllers/user_controller.php [hunk 30]
```diff
@@ -91,29 +112,137 @@ public function profile()
...
+        switch ($page) {
+            case 'change_password':
+                break;
+            case 'edit_success':
+                $user->old_password = Param::get('old_password');
+                $user->password = Param::get('password');
+                $user->confirm_password = Param::get('confirm_password');
+                        
+                try {
+                    $change_password = $user->changePassword();
```
Comment (fatimaPHP) [inline app/controllers/user_controller.php]: please add restriction po if the current user is the one changing his own password :bow:

File: app/controllers/comment_controller.php (hunk 1 above)
Comment (dandiesta) [inline app/controllers/comment_controller.php:101]: done po.

File: app/controllers/comment_controller.php (hunk 3 above)
Comment (dandiesta) [inline app/controllers/comment_controller.php]: done na po.

File: app/models/user.php (hunk 7 above)
Comment (dandiesta) [inline app/models/user.php]: done po.

File: app/models/user.php (hunk 8 above)
Comment (dandiesta) [inline app/models/user.php]: done po

File: app/views/layouts/default.php (hunk 10 above)
Comment (dandiesta) [inline app/views/layouts/default.php]: I have edited some designs in bootstrap. but it doesn't apply unless i change bootstrap.css itself. that's why I included that po.

File: app/app_controller.php (hunk 13 above)
Comment (dandiesta) [inline app/app_controller.php]: done na po

Comment (dandiesta) [inline app/app_controller.php]: I only used == instead of ===. There is an error if I use ===

File: app/models/thread.php (hunk 26 above)
Comment (dandiesta) [inline app/models/thread.php]: done po

File: app/models/user.php (hunk 28 above)
Comment (dandiesta) [inline app/models/user.php]: noted po. :)

File: app/app_controller.php (hunk 29 above)
Comment (dandiesta) [inline app/app_controller.php:30]: yes po

File: app/models/user.php (hunk 27 above)
Comment (dandiesta) [inline app/models/user.php]: noted po

File: app/models/comment.php (hunk 24 above)
Comment (dandiesta) [inline app/models/comment.php:220]: noted po

File: app/models/user.php (hunk 23 above)
Comment (dandiesta) [inline app/models/user.php]: noted po.

File: app/controllers/thread_controller.php (hunk 19 above)
Comment (dandiesta) [inline app/controllers/thread_controller.php]: done po

File: vendor/SimplePagination.php (hunk 20 above)
Comment (dandiesta) [inline vendor/SimplePagination.php]: noted po

File: app/controllers/comment_controller.php (hunk 15 above)
Comment (dandiesta) [inline app/controllers/comment_controller.php]: done po

File: .gitignore (hunk 12 above)
Comment (dandiesta) [inline .gitignore]: done po

File: app/models/user.php (hunk 27 above)
Comment (dandiesta) [inline app/models/user.php]: nameChecker has been moved to a helper but the passwordChecker remains in the model. When I tried to move passwordChecker, an error occured

File: app/controllers/comment_controller.php (hunk 16 above)
Comment (dandiesta) [inline app/controllers/comment_controller.php]: done po

File: app/models/comment.php (hunk 25 above)
Comment (dandiesta) [inline app/models/comment.php]: done po

File: app/views/user/profile.php (hunk 11 above)
Comment (dandiesta) [inline app/views/user/profile.php]: done po

File: app/controllers/comment_controller.php (hunk 14 above)
Comment (dandiesta) [inline app/controllers/comment_controller.php]: done po

File: app/controllers/comment_controller.php (hunk 17 above)
Comment (dandiesta) [inline app/controllers/comment_controller.php:197]: if this method becomes static, an error occurs :bow:

File: app/controllers/comment_controller.php (hunk 18 above)
Comment (dandiesta) [inline app/controllers/comment_controller.php:208]: noted po

File: app/controllers/comment_controller.php (hunk 21 above)
Comment (dandiesta) [inline app/controllers/comment_controller.php]: i just removed the constant inside the library po then defined it na lang po sa core

File: app/controllers/user_controller.php (hunk 22 above)
Comment (dandiesta) [inline app/controllers/user_controller.php]: done po

File: app/controllers/user_controller.php (hunk 30 above)
Comment (dandiesta) [inline app/controllers/user_controller.php]: done
